agent/structs/config_entry_routes.go

```diff
@@ -0,0 +1,177 @@
+type HTTPRouteConfigEntry struct {
+	// Kind of the config entry. This will be set to structs.HTTPRoute.
+	Kind string
+
+	// Name is used to match the config entry with its associated set
+	// of resources, which may include routers, splitters, filters, etc.
+	Name string
+
+	Meta               map[string]string `json:",omitempty"`
+	acl.EnterpriseMeta `hcl:",squash" mapstructure:",squash"`
+	RaftIndex
+}
+
+func (e *HTTPRouteConfigEntry) GetKind() string {
+	return HTTPRoute
+}
+
+func (e *HTTPRouteConfigEntry) GetName() string {
+	if e == nil {
+		return ""
+	}
+	return e.Name
+}
+
+func (e *HTTPRouteConfigEntry) Normalize() error {
+	return nil
+}
+
+func (e *HTTPRouteConfigEntry) Validate() error {
+	return nil
+}
+
+func (e *HTTPRouteConfigEntry) CanRead(authz acl.Authorizer) error {
+	var authzContext acl.AuthorizerContext
+	e.FillAuthzContext(&authzContext)
+	return authz.ToAllowAuthorizer().MeshReadAllowed(&authzContext)
+}
+
+func (e *HTTPRouteConfigEntry) CanWrite(authz acl.Authorizer) error {
+	var authzContext acl.AuthorizerContext
+	e.FillAuthzContext(&authzContext)
+	return authz.ToAllowAuthorizer().MeshWriteAllowed(&authzContext)
+}
+
+func (e *HTTPRouteConfigEntry) GetMeta() map[string]string {
+	if e == nil {
+		return nil
+	}
+	return e.Meta
+}
+
+func (e *HTTPRouteConfigEntry) GetEnterpriseMeta() *acl.EnterpriseMeta {
+	if e == nil {
+		return nil
+	}
+	return &e.EnterpriseMeta
+}
+
+func (e *HTTPRouteConfigEntry) GetRaftIndex() *RaftIndex {
+	if e == nil {
+		return &RaftIndex{}
+	}
+	return &e.RaftIndex
+}
+
+// TCPRouteConfigEntry manages the configuration for a TCP route
+// with the given name.
+type TCPRouteConfigEntry struct {
+	// Kind of the config entry. This will be set to structs.TCPRoute.
+	Kind string
+
+	// Name is used to match the config entry with its associated set
+	// of resources.
+	Name string
+
+	// Parents is a list of gateways that this route should be bound to
+	Parents []ResourceReference
+	// Services is a list of TCP-based services that this should route to.
+	// Currently, this must specify at maximum one service.
+	Services []TCPService
+
+	Meta map[string]string `json:",omitempty"`
+	// Status is the asynchronous status which a TCPRoute propagates to the user.
+	Status             Status
+	acl.EnterpriseMeta `hcl:",squash" mapstructure:",squash"`
+	RaftIndex
+}
+
+func (e *TCPRouteConfigEntry) GetKind() string {
+	return TCPRoute
+}
+
+func (e *TCPRouteConfigEntry) GetName() string {
+	if e == nil {
+		return ""
+	}
+	return e.Name
+}
+
+func (e *TCPRouteConfigEntry) GetMeta() map[string]string {
+	if e == nil {
+		return nil
+	}
+	return e.Meta
+}
+
+func (e *TCPRouteConfigEntry) Normalize() error {
+	for i, parent := range e.Parents {
+		if parent.Kind == "" {
+			parent.Kind = APIGateway
+			e.Parents[i] = parent
+		}
+	}
+	return nil
+}
+
+func (e *TCPRouteConfigEntry) Validate() error {
+	validParentKinds := map[string]bool{
+		APIGateway: true,
+	}
+
+	if len(e.Services) > 1 {
```

Comment:
Is this temporary. It looks like the TCPService struct has a weight which would give the impression of allowing multiple services and traffic splitting.

Reply:
Yes, we don't currently support multiple TCP services bound to a route or multiple routes bound to a gateway listener. The spec allows for it. There was some discussion about changing this approach in the future, but currently the thought is that load balancing between TCP connections might not make sense to users because the connections are persistent.

Reply:
I thought this was just a limitation of the current ingress gateway implementation under the hood, and something that we likely _will_ want to support once we're generating our own xDS directly.